<a href="https://colab.research.google.com/github/Reign2121/Mining-project2/blob/main/%EA%B3%B5%EB%A6%89_%ED%9A%8C%EA%B7%80.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

회귀분석입니다! 정확히는 다중회귀분석(or 다변량)이라고 합니다.

회귀분석의 모델을 검정하고 평가하는 방법은 크게 두 가지 방향이 있습니다.

1. 통계적 관점, t-test로 계수들의 유의성을 검정하고, f-test로 모델의 유의성을 검정합니다.
2. CS 관점, 대량 데이터를 통한 test 결과에 따라 유의성을 검정합니다.

저는 이 프로젝트에서 데이터의 수가 충분하지 않고, 지나치게 많고 복잡한 변수들을 문제점으로 식별하고 통계적인 스탠스를 취했습니다.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
import numpy as np
G = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/Colab/머신러닝 입문/Data/공릉동_상권분석_최최종_.csv",encoding = 'cp949')
G = pd.get_dummies(G)
G.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   분기당_매출_금액            506 non-null    int64
 1   분기당_매출_건수            506 non-null    int64
 2   남성_매출_비율             506 non-null    int64
 3   여성_매출_비율             506 non-null    int64
 4   연령대_10_매출_비율         506 non-null    int64
 5   연령대_20_매출_비율         506 non-null    int64
 6   연령대_30_매출_비율         506 non-null    int64
 7   서비스_업종_코드_명_농수산물     506 non-null    uint8
 8   서비스_업종_코드_명_마트/편의점   506 non-null    uint8
 9   서비스_업종_코드_명_미용       506 non-null    uint8
 10  서비스_업종_코드_명_생활       506 non-null    uint8
 11  서비스_업종_코드_명_서적/문구    506 non-null    uint8
 12  서비스_업종_코드_명_스포츠      506 non-null    uint8
 13  서비스_업종_코드_명_여가       506 non-null    uint8
 14  서비스_업종_코드_명_음식점      506 non-null    uint8
 15  서비스_업종_코드_명_의류/신발    506 non-null    uint8
 16  서비스_업종_코드_명_의원       506 n

statsmodel 결과를 보고 t-test 통과한 변수들만 남기기!!

In [ ]:
y = G['분기당_매출_건수']
X1 = G.iloc[:,0:20]
X1 = X1.drop('분기당_매출_건수', axis = 1)
X1 = X1.drop('서비스_업종_코드_명_호프-간이주점', axis = 1)
X1 = X1.drop('서비스_업종_코드_명_의류/신발', axis = 1)
X1 = X1.drop('서비스_업종_코드_명_생활', axis = 1)
X1 = X1.drop('서비스_업종_코드_명_서적/문구', axis = 1)
X1 = X1.drop('서비스_업종_코드_명_스포츠', axis = 1)
X1 = X1.drop('서비스_업종_코드_명_학원', axis = 1)
X1 = X1.drop('서비스_업종_코드_명_농수산물', axis = 1)
X1 = X1.drop('연령대_30_매출_비율', axis = 1)
X1 = X1.drop('남성_매출_비율', axis = 1)
X1 = X1.drop('여성_매출_비율', axis = 1)

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X1,y, test_size = 0.2, random_state = 42)

In [ ]:
from sklearn.linear_model import LinearRegression
lr_model = LinearRegression()
lr_model.fit(X_train, y_train)
y_predict = lr_model.predict(X_test) #모형 학습에는 전혀 쓰이지 않은 테스트 데이터를 넣어주어야 한다
np.set_printoptions(suppress=True, precision=3)
print("회귀계수:", lr_model.coef_, "상수항:", lr_model.intercept_)

회귀계수: [     0.     -7673.347    937.579 239842.014 -33824.04   19356.384
  16713.532  20768.943  95247.744] 상수항: -11541.849515143978


In [ ]:
np.set_printoptions(suppress=True, precision=3)

In [ ]:
R2 = lr_model.score(X_train, y_train)
print("결정계수:", R2)

결정계수: 0.7881874424454021


In [ ]:
import statsmodels.api as sm 
results = sm.OLS(y, sm.add_constant(X1)).fit() 
results.summary()

/usr/local/lib/python3.7/dist-packages/statsmodels/tsa/tsatools.py:117: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              분기당_매출_건수   R-squared:                       0.802
Model:                            OLS   Adj. R-squared:                  0.799
Method:                 Least Squares   F-statistic:                     223.6
Date:                Mon, 23 May 2022   Prob (F-statistic):          2.55e-168
Time:                        14:36:19   Log-Likelihood:                -6129.8
No. Observations:                 506   AIC:                         1.228e+04
Df Residuals:                     496   BIC:                         1.232e+04
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const              -1.145e+04   3305.888     -3.464      0.001   -1.79e+04   -4956.951
분기당_매출_금액           3.335e-05   1.27e-06     26.236      0.000    3.09e-05    3.58e-05
연령대_10_매출_비율       -6997.1621   1554.715     -4.501      0.000   -1.01e+04   -3942.522
연령대_20_매출_비율         898.4996    148.683      6.043      0.000     606.374    1190.625
서비스_업종_코드_명_마트/편의점  2.415e+05   1.05e+04     23.016      0.000    2.21e+05    2.62e+05
서비스_업종_코드_명_미용     -2.554e+04   7115.888     -3.589      0.000   -3.95e+04   -1.16e+04
서비스_업종_코드_명_여가      1.948e+04   8898.132      2.190      0.029    2000.277     3.7e+04
서비스_업종_코드_명_음식점     1.533e+04   6123.390      2.503      0.013    3296.964    2.74e+04
서비스_업종_코드_명_의원      2.709e+04   7192.007      3.766      0.000     1.3e+04    4.12e+04
서비스_업종_코드_명_제과/음료    8.45e+04      1e+04      8.415      0.000    6.48e+04    1.04e+05
==============================================================================
Omnibus:                       82.671   Durbin-Watson:                   2.225
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              939.770
Skew:                           0.234   Prob(JB):                    8.54e-205
Kurtosis:                       9.660   Cond. No.                     1.20e+10
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.2e+10. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

통계적으로 유의미한 모델 도출 성공!!

adj r^2 = 0.799


F의 P -value / T의 P-value 모두 유의수준 0.05보다 작습니다!

확실히 대학가 이기 때문에 20대의 매출비율이 상대적으로 높네요.

미용분야의 업종이 저조하고 제과/음료가 매출 건수에 가장 큰 영향력을 미치고 있습니다.

매출건수에 미치는 영향력들을 고려하여 마케팅 전략을 도출할 수 있겠죠!!